intern/cycles/kernel/kernel_color.h

```diff
@@ -33,6 +33,673 @@ ccl_device float linear_rgb_to_gray(KernelGlobals *kg, float3 c)
   return dot(c, float4_to_float3(kernel_data.film.rgb_to_y));
 }
 
+ccl_device float float_lerp(float start, float end, float progress)
+{
+    return start + ((end - start) * progress);
+}
+
+ccl_device float3 find_position_in_lookup_unit_step(float4 lookup[], float position_to_find, int start, int end)
+{
+    if(position_to_find <= start) {
+        return make_float3(
+            lookup[0][1],
+            lookup[0][2],
+            lookup[0][3]
+        );
+    }
+    if(position_to_find >= end) {
+        return make_float3(
+            lookup[end][1],
+            lookup[end][2],
+            lookup[end][3]
+        );
+    }
+    int lower_bound = int(position_to_find) - start;
+    int upper_bound = lower_bound + 1;
+    float progress = position_to_find - int(position_to_find);
+    return make_float3(
+        float_lerp(lookup[lower_bound][1], lookup[upper_bound][1], progress),
+        float_lerp(lookup[lower_bound][2], lookup[upper_bound][2], progress),
+        float_lerp(lookup[lower_bound][3], lookup[upper_bound][3], progress)
+    );
+}
+
+ccl_device float3 wavelength_to_xyz(float wavelength)
+{
+    float4 wavelength_xyz_lookup[] = {
```

Comment:
This is a lookup table with the values of the CIE XYZ coordinates of each wavelength. It is used when taking spectra back into RGB.

While this is really bad for performance I couldn't figure out how to get this working otherwise, I think it has to do with the `ccl_device` but I don't fully understand the implications of that either, I _think_ that specifies that this should run on the GPU if available.

If we can find a nice approach to load things into `KernelGlobals` kg at startup, that would make many things less hardcoded and more flexible.